v2/workloadapi/client_test.go

```diff
@@ -1 +1,210 @@
 package workloadapi
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/spiffe/go-spiffe/v2/bundle/jwtbundle"
+	"github.com/spiffe/go-spiffe/v2/proto/spiffe/workload"
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+	"github.com/spiffe/go-spiffe/v2/svid/jwtsvid"
+	"github.com/spiffe/go-spiffe/v2/test"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	"gopkg.in/square/go-jose.v2"
+)
+
+type ClientTestSuite struct {
+	suite.Suite
+	ca     *test.CA
+	wl     *test.WorkloadAPI
+	client *Client
+}
+
+func TestClientTestSuite(t *testing.T) {
+	suite.Run(t, new(ClientTestSuite))
+}
+
+func (s *ClientTestSuite) SetupTest() {
+	s.ca = test.NewCA(s.T())
+	s.wl = test.NewWorkloadAPI(s.T())
+	resp := s.makeX509SVIDResponse([]string{"spiffe://example.org/foo", "spiffe://example.org/bar"})
+	s.wl.SetX509SVIDResponse(resp)
+
+	spiffeID := "spiffe://example.org/mytoken"
+	audience := []string{"spiffe://example.org/target", "spiffe://example.org/extraAudience"}
+	respJWT := s.makeJWTSVIDResponse(spiffeID, audience)
+	_ = s.makeJWTSVIDResponse(spiffeID, audience)
+	s.wl.SetJWTSVIDResponse(respJWT)
+
+	jwk1 := s.ca.CreateJWK("1")
+	jwk2 := s.ca.CreateJWK("2")
+	s.wl.SetJWTBundle("spiffe://example.org", []jose.JSONWebKey{jwk1, jwk2})
+
+	s.client, _ = New(context.TODO(), WithAddr(s.wl.Addr()))
+}
+
+func (s *ClientTestSuite) TearDownTest() {
+	s.client.Close()
+	s.wl.Stop()
+}
+
+func (s *ClientTestSuite) TestFetchX509SVID() {
+	svid, err := s.client.FetchX509SVID(context.TODO())
+	require.NotEmpty(s.T(), svid)
+	require.Nil(s.T(), err)
+}
+
+func (s *ClientTestSuite) TestFetchX509SVIDs() {
+	svid, err := s.client.FetchX509SVIDs(context.TODO())
+	require.Len(s.T(), svid, 2)
+	require.Nil(s.T(), err)
+}
+
+func (s *ClientTestSuite) TestFetchX509Bundles() {
+	svid, err := s.client.FetchX509Bundles(context.TODO())
+	require.NotEmpty(s.T(), svid)
+	require.Nil(s.T(), err)
+}
+
+func (s *ClientTestSuite) TestFetchX509Context() {
+	svid, err := s.client.FetchX509Context(context.TODO())
+	require.Nil(s.T(), err)
+	require.NotEmpty(s.T(), svid)
+	require.Len(s.T(), svid.SVIDs, 2)
+}
+
+func (s *ClientTestSuite) TestWatchX509Context() {
+	w := newTestWatcher(s.T())
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	go func() {
+		_ = s.client.WatchX509Context(ctx, w)
```

Comment:
We should wait for this goroutine to go away before allowing the test function to return (via waitgroup?)